---
title: "Phylogenetic pipeline"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

We will outline an example pipeline using external command-line tools, called
through `outsider`.

In this example, we will generate a phylogenetic tree in three steps:
obtain orthologous sequences, align the sequences, estimate tree.

In our example, we will keep it simple and fast and generate a tree of CytB 
sequences in the primate genus of Night Monkeys, **Aotus**.

## phylotaR

To get us started with need DNA sequences that represent the same gene region,
i.e. sequences that are orthologous. We can obtain our sequences quickly using
the example data from the R package [`phylotaR`](https://github.com/ropensci/phylotaR).

`phylotaR` provides a pipeline for identifying orthologous sequences from a
reputable online biological sequence database. The package comes with a few
pre-calculated example sequences. We can extract sequence clusters that
represent the gene region [cytb](https://en.wikipedia.org/wiki/Cytochrome_b)
with the following script.

```{r phylotar}
library(phylotaR)
data("aotus")
smmry <- summary(aotus)
cytb <- smmry$ID[which(grepl('cytb', smmry$Feature))[1]]
cytb <- drop_clstrs(aotus, cytb)
cytb <- drop_by_rank(cytb, n = 1)
txids <- get_txids(cytb, cid = '9')
sp_nms <- get_tx_slot(cytb, txids, slt_nm = 'scnm')
sp_nms <- sub(' ', '_', sp_nms)
write_sqs(cytb, 'aotus_cytb.fasta', sq_nm = sp_nms, sid = names(sp_nms))
```

## Alignment

Now we have written our sequences to a text-based `.fasta` file, we need to
aign them! To do this, we can use
[`mafft`](https://mafft.cbrc.jp/alignment/software/). We first need to install
the module and then import the key function, `mafft`. After doing this, we can
call the `mafft` program using the same arguments that we would if it were
calling the program via command-line.

```{r mafft}
library(outsider)
repo <- 'dombennett/om..mafft'
suppressWarnings(module_install(repo = repo))
mafft <- module_import(fname = 'mafft', repo = repo)
mafft('--auto', 'aotus_cytb.fasta', '>', 'alignment.fasta')
```
## Phylogeny

We can repeat the same process as we did for `mafft` but instead for
[`RAxML`](https://github.com/stamatak/standard-RAxML) -- a command-line program
for estimating phylogenetic trees using maximum likelihood.

```{r raxml}
library(outsider)
repo <- 'dombennett/om..raxml'
suppressWarnings(module_install(repo = repo))
raxml <- module_import(fname = 'raxml', repo = repo)
raxml('-m', 'GTRGAMMA', '-s', 'alignment.fasta', '-p', '1234', '-n',
      'aotus_cytb', '-T', '2')
```

## Visualisation

Now, let's check out our tree!

```{r visualise}
library(ape)
tree <- read.tree('RAxML_bestTree.aotus_cytb')
plot(tree, type = 'unrooted')
```

```{r cleanup, include = FALSE}
files_to_remove <- c(list.files(pattern = c('.fasta')),
                     list.files(pattern = 'RAxML_'))
for (file_to_remove in files_to_remove) {
  try(file.remove(file_to_remove), silent = TRUE)
}
repos <- c('dombennett/om..raxml', 'dombennett/om..mafft')
for (repo in repos) {
  try(module_uninstall(repo), silent = TRUE)
}
```